packages/compass-generative-ai/src/tools-controller.ts

```diff
@@ -1,25 +1,126 @@
+  TransportRunnerBase,
+  UserConfigSchema,
+} from 'mongodb-mcp-server';
+import type {
+  Server,
+  TransportRunnerConfig,
+  UserConfig,
+} from 'mongodb-mcp-server';
+import { AllTools } from 'mongodb-mcp-server/tools';
+import { createConnectionErrorHandler } from './tools-connection-error-handler';
+import { ToolsLogger } from './tools-logger';
+import { ToolsConnectionManager } from './tools-connection-manager';
+import type { ToolsConnectParams } from './tools-connection-manager';
 
-// TODO: add readonly-db
-type ToolGroup = 'compass';
+type ToolGroup = 'compass' | 'db-read';
 
 type CompassContext = {
   query?: string;
   aggregation?: string;
 };
 
-// TODO: add connection info
-type ToolsContext = CompassContext;
+type ToolsContext = CompassContext & {
+  connection?: ToolsConnectParams;
+};
+
+// TODO: rather than string there must be a better type we can use here
+const readonlyTools = new Set<string>([
+  'find',
+  'aggregate',
+  'list-databases',
+  'count',
+  'list-databases',
+  'list-collections',
+  'collection-indexes',
+  'collection-schema',
+  //'explain',
+  // TODO: we're only allowd 10 tools at the moment
+  //'collection-storage-size',
+  //'db-stats',
+  //'mongodb-logs',
+]);
+
+/**
+ * In-memory MCP runner that doesn't bind to any external transport.
+ * The server is kept in memory for direct programmatic access.
+ */
+class InMemoryRunner extends TransportRunnerBase {
+  public server?: Server;
+  public readonly userConfig: UserConfig;
+
+  constructor(config: TransportRunnerConfig) {
+    super(config);
+    this.userConfig = config.userConfig;
+  }
+
+  async start(): Promise<void> {
+    this.server = await this.setupServer();
+  }
+
+  async closeTransport(): Promise<void> {
+    await this.server?.close();
+  }
+}
+
+type ToolsControllerConfig = {
+  logger: Logger;
+  getTelemetryAnonymousId: () => string;
+};
 
 export class ToolsController {
   private logger: Logger;
   private toolGroups: Set<ToolGroup> = new Set();
   private context: ToolsContext = Object.create(null);
+  private readonly runner: InMemoryRunner;
+  private connectionManager: ToolsConnectionManager;
+  private currentConnection?: ToolsConnectParams;
 
-  constructor({ logger }: { logger: Logger }) {
+  constructor({ logger, getTelemetryAnonymousId }: ToolsControllerConfig) {
     this.logger = logger;
+    const mcpConfig = UserConfigSchema.parse({
+      disabledTools: ['connect'],
+      loggers: ['mcp'],
+    });
+
+    this.runner = new InMemoryRunner({
+      userConfig: mcpConfig,
+      createConnectionManager: () => Promise.resolve(this.connectionManager),
+      connectionErrorHandler: createConnectionErrorHandler(),
+      additionalLoggers: [new ToolsLogger(Keychain.root, this.logger)],
+      telemetryProperties: {
+        hosting_mode: 'compass',
+      },
+    });
+
+    this.connectionManager = new ToolsConnectionManager({
+      // TODO: this is a hack so that we can have the connection manager always
```

Comment:
The way mongodb-mcp-server seems to be architected is that the connection manager only gets created asynchronously. (see createConnectionManager) above. But then the object here becomes possibly undefined which is a pain.

Ideally I'd like it to always exist and just start off disconnected. Simplifies code a bit.

I'm assuming here that the connection manager's logger is the same logger that the runner uses.